### PR TITLE
Add vanilla loss to NPO

### PR DESCRIPTION
**Pull request by CatherineSue**
1. VPG uses a different loss function than NPO and its variants. VPG
uses the vanilla loss, as VPG uses direct gradient estimator from policy
score function, NPO calculates the derivation using importance sampling.
Thus, we need to add the 'real' vanilla loss to NPO to inherit from NPO.
The Enum is kept in this commit instead of string.
2. Update examples.

### CODE REVIEW

**Comment by CatherineSue** [comment]
Regression of tf/VPG vs. theano/VPG.
screen shot 2018-12-20 at 5 14 18 pm

**Comment by ryanjulian** [comment]
isn't PenaltyLbfgsOptimizer TRPO? Because the penalty term is for KL-divergence?

File: garage/tf/algos/npo.py [hunk 1]
```diff
@@ -17,18 +17,19 @@
...
 from garage.tf.misc.tensor_utils import graph_inputs
-from garage.tf.optimizers import LbfgsOptimizer
+from garage.tf.optimizers import PenaltyLbfgsOptimizer
 
 
 @unique
 class PGLoss(Enum):
     VANILLA = "vanilla"
-    CLIP = "clip"
+    SURROGATE_VANILLA = "surrogate_vanilla"
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: I think just `SURROGATE`/`"surrogate"` is fine here

File: garage/tf/algos/npo.py [hunk 2]
```diff
@@ -300,6 +301,10 @@ def _build_policy_loss(self, i):
                         policy_dist_info,
                         name="lr")
 
+                    ll = pol_dist.log_likelihood_sym(
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: seems like the scope above should now just be "loss" since it's not necessarily surrogate

File: garage/tf/algos/npo.py [hunk 3]
```diff
@@ -308,12 +313,21 @@ def _build_policy_loss(self, i):
...
+                        policy_dist_info_valid,
+                        name="log_likelihood")
+
+                    vanilla = ll * adv_valid
                     surr_vanilla = lr * adv_valid
 
                 if self._pg_loss == PGLoss.VANILLA:
-                    # VPG, TRPO use the standard surrogate objective
+                    # VPG uses the vanilla objective
+                    surr_obj = tf.identity(vanilla, name="surr_obj")
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: i think the variable `surr_obj` now needs to be renamed since it's not necessarily surrogate

File: garage/tf/algos/npo.py [hunk 4]
```diff
@@ -308,12 +313,21 @@ def _build_policy_loss(self, i):
                         policy_dist_info_valid,
                         name="lr")
 
+                    ll = pol_dist.log_likelihood_sym(
+                        i.valid.action_var,
+                        policy_dist_info_valid,
+                        name="log_likelihood")
+
+                    vanilla = ll * adv_valid
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: I think it might be more clear if you moved `vanilla` and `surr_vanilla` inside the `if` statement, even if it means copying a line or two.

File: garage/tf/algos/npo.py [hunk 5]
```diff
@@ -300,6 +301,10 @@ def _build_policy_loss(self, i):
                         policy_dist_info,
                         name="lr")
 
+                    ll = pol_dist.log_likelihood_sym(
+                        i.action_var, policy_dist_info, name="log_likelihood")
+
+                    vanilla = ll * advantages * i.valid_var
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: i think it might be more clear if you moved these inside the `if` statement, even if it means copying a couple lines

File: garage/tf/algos/npo.py [hunk 6]
```diff
@@ -17,18 +17,19 @@
...
-    CLIP = "clip"
+    SURROGATE_VANILLA = "surrogate_vanilla"
+    SURROGATE_CLIP = "surrogate_clip"
 
 
 class NPO(BatchPolopt):
     def __init__(self,
-                 pg_loss=PGLoss.VANILLA,
+                 pg_loss=PGLoss.SURROGATE_VANILLA,
                  clip_range=0.01,
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: while we are here, i think we should separate the `clip_range` and `step_size` parameters, since they are different numbers in the literature.

`step_size` refers to the maximum KL divergence between old and new policies, as in TRPO
`clip_range` refers to the limit on the likelihood ratio between policies, as in PPO.

File: garage/tf/algos/npo.py [hunk 7]
```diff
@@ -49,7 +50,7 @@ def __init__(self,
         if optimizer is None:
             if optimizer_args is None:
                 optimizer_args = dict()
-            optimizer = LbfgsOptimizer
+            optimizer = PenaltyLbfgsOptimizer
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: I think this is an error. True VPG is unconstrainted in KL divergence. Using `PenaltyLbfgsOptimizer` turns it into the soft-penalty version of TRPO.

@amolchanov86 do you agree?

Comment (amolchanov86) [reply garage/tf/algos/npo.py]: you are right

File: garage/tf/algos/npo.py (hunk 6 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Is the `step_size` same as the `epsilon` in f(kl) < epsilon?

File: garage/tf/algos/npo.py (hunk 1 above)
Comment (amolchanov86) [reply garage/tf/algos/npo.py]: The shorter the better

File: garage/tf/algos/npo.py (hunk 6 above)
Comment (amolchanov86) [reply garage/tf/algos/npo.py]: I would then explicitly rename them for what they are `kl_max_step` , `loss_clip_range`

**Comment by CatherineSue** [comment]
This is accidentally closed because of a wrong git reset cmd, I've pushed the new changes.

File: garage/tf/algos/npo.py [hunk 8]
```diff
@@ -310,10 +326,15 @@ def _build_policy_loss(self, i):
 
                     surr_vanilla = lr * adv_valid
 
+            # Finalize objective function
+            with tf.name_scope("final_loss"):
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: just "loss"

File: garage/tf/algos/npo.py [hunk 9]
```diff
@@ -310,10 +326,15 @@ def _build_policy_loss(self, i):
 
                     surr_vanilla = lr * adv_valid
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: "surrogate", to be consistent with the enum

File: garage/tf/algos/npo.py [hunk 10]
```diff
@@ -291,6 +292,21 @@ def _build_policy_loss(self, i):
...
+                else:
+                    ll = pol_dist.log_likelihood_sym(
+                        i.valid.action_var,
+                        policy_dist_info_valid,
+                        name="log_likelihood")
+
+                    vanilla = ll * adv_valid
+
             # Calculate surrogate loss
             with tf.name_scope("surr_loss"):
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: "surrogate_loss"

File: garage/tf/algos/npo.py [hunk 11]
```diff
@@ -323,18 +344,16 @@ def _build_policy_loss(self, i):
...
+                    obj = tf.minimum(surr_vanilla, surr_clip, name="surr_obj")
                 else:
                     raise NotImplementedError("Unknown PGLoss")
 
                 # Maximize E[surrogate objective] by minimizing
                 # -E_t[surrogate objective]
                 if self.policy.recurrent:
-                    surr_loss = (-tf.reduce_sum(surr_obj)) / tf.reduce_sum(
-                        i.valid_var)
+                    loss = (-tf.reduce_sum(obj)) / tf.reduce_sum(i.valid_var)
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: unnecessary ()

File: garage/tf/algos/npo.py [hunk 12]
```diff
@@ -323,18 +344,16 @@ def _build_policy_loss(self, i):
                         surr_clip = lr_clip * advantages * i.valid_var
                     else:
                         surr_clip = lr_clip * adv_valid
-                    surr_obj = tf.minimum(
-                        surr_vanilla, surr_clip, name="surr_obj")
+                    obj = tf.minimum(surr_vanilla, surr_clip, name="surr_obj")
                 else:
                     raise NotImplementedError("Unknown PGLoss")
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: can you test this case?

File: garage/tf/algos/npo.py [hunk 13]
```diff
@@ -323,18 +344,16 @@ def _build_policy_loss(self, i):
                         surr_clip = lr_clip * advantages * i.valid_var
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: it seems that the recurrent case here is never tested. can you add a test which uses recurrent too, so that this line is covered?

File: garage/tf/algos/npo.py (hunk 12 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Tested. Also tested "Unknown KLConstraint"

File: garage/tf/algos/npo.py (hunk 13 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Added in tests/garage/tf/algos/test_ppo.py.

File: garage/tf/algos/npo.py (hunk 11 above)
Comment (CatherineSue) [reply garage/tf/algos/npo.py]: Deleted

File: setup.cfg [hunk 14]
```diff
@@ -10,6 +10,8 @@ import-order-style = google
 application-import-names = tests,sandbox,garage,examples
 
 [pylint]
+#[MASTER]
+init-hook='import sys; sys.path.append("./")'
```
Comment (ryanjulian) [inline setup.cfg]: did this slip in on accident?

Comment (CatherineSue) [reply setup.cfg]: Oh Yes. I'll delete it now.

File: garage/tf/algos/npo.py [hunk 15]
```diff
@@ -23,13 +24,28 @@
...
+    SURROGATE_CLIP = "surrogate_clip"
 
 
 class NPO(BatchPolopt):
+    """
+    Natural Policy Gradient Optimization.
+
+    Attributes:
+        name(str): The name of the algorithm.
+        loss_clip_range(float): The limit on the likelihood ratio between
```
Comment (ryanjulian) [inline garage/tf/algos/npo.py]: lr_clip_range

File: examples/tf/ppo_pendulum.py [hunk 16]
```diff
@@ -39,7 +39,7 @@ def run_task(*_):
         max_path_length=100,
         n_itr=488,
         discount=0.99,
-        step_size=0.01,
+        loss_clip_range=0.01,
```
Comment (CatherineSue) [inline examples/tf/ppo_pendulum.py]: I've separated `loss_clip_range` and `max_kl_step`. Theano/NPO still uses `step_size`, but I think we can leave it there. 
There is a problem when I change the corresponding params in tests/benchmarks/test_benchmark_*.py. The pylint raises the following errors:
```
hookid: pylint

************* Module tests.benchmarks.test_benchmark_ppo
tests/benchmarks/test_benchmark_ppo.py:E0401: 14,0: Unable to import 'baselines.bench'
tests/benchmarks/test_benchmark_ppo.py:E0401: 16,0: Unable to import 'baselines.bench'
tests/benchmarks/test_benchmark_ppo.py:E0401: 18,0: Unable to import 'baselines.common'
tests/benchmarks/test_benchmark_ppo.py:E0401: 19,0: Unable to import 'baselines.common.vec_env.dummy_vec_env'
tests/benchmarks/test_benchmark_ppo.py:E0401: 20,0: Unable to import 'baselines.common.vec_env.vec_normalize'
tests/benchmarks/test_benchmark_ppo.py:E0401: 21,0: Unable to import 'baselines.logger'
tests/benchmarks/test_benchmark_ppo.py:E0401: 22,0: Unable to import 'baselines.ppo2'
tests/benchmarks/test_benchmark_ppo.py:E0401: 23,0: Unable to import 'baselines.ppo2.policies'
tests/benchmarks/test_benchmark_ppo.py:W0611: 17,0: Unused common imported from baselines
************* Module tests.benchmarks.test_benchmark_trpo
tests/benchmarks/test_benchmark_trpo.py:E0401: 15,0: Unable to import 'baselines.bench'
tests/benchmarks/test_benchmark_trpo.py:E0401: 16,0: Unable to import 'baselines.common.cmd_util'
tests/benchmarks/test_benchmark_trpo.py:E0401: 17,0: Unable to import 'baselines.common.tf_util'
tests/benchmarks/test_benchmark_trpo.py:E0401: 18,0: Unable to import 'baselines.ppo1.mlp_policy'
tests/benchmarks/test_benchmark_trpo.py:E0401: 19,0: Unable to import 'baselines.trpo_mpi'
```
Anybody have ideas why?

Comment (ryanjulian) [reply examples/tf/ppo_pendulum.py]: this is because pylint is now checking whether all of your imports are actually importable from the default garage install.

are you sure that those baselines modules exist in the default environment.yml? the packages changed and it needs updating?

File: garage/tf/algos/ppo.py [hunk 17]
```diff
@@ -24,7 +24,7 @@ def __init__(self, optimizer=None, optimizer_args=None, **kwargs):
             if optimizer_args is None:
                 optimizer_args = dict()
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply garage/tf/algos/ppo.py]: no i think this is okay

File: garage/tf/algos/npo.py (hunk 7 above)
Comment (naeioi) [reply garage/tf/algos/npo.py]: This is probably not an error. This is not relevant to VPG because VPG overrides optimizer to FirstOrderOptimizer. I think Chang was to sync up TF with Theano. https://github.com/rlworkgroup/garage/blob/241ae5c02f17e6fb4c6ec59b35454fe4b3dd1c23/garage/theano/algos/npo.py#L26

NPO also uses interface unique of PenaltyLbfgsOptimizer. https://github.com/rlworkgroup/garage/blob/241ae5c02f17e6fb4c6ec59b35454fe4b3dd1c23/garage/theano/algos/npo.py#L93-L99

Comment (ryanjulian) [reply garage/tf/algos/npo.py]: ah i see. so it doesn't matter.

i guess Artem and I just assumed that VPG would be the "default" version of NPO.

Comment (ryanjulian) [reply garage/tf/algos/npo.py]: I guess in the end it doesn't matter because nobody runs NPO directly, and if an algorithm uses a specific optimizer it will set it in the constructor.

I'm ambivalent about whether to revert it or not. Whatever Chang wants.

**Comment by ryanjulian** [comment]
Anything blocking merging this?
